File: DFS&BFS/javascript/9019.js
const input = require("fs").readFileSync("/dev/stdin").toString().split("\n");

const cal = {
  D: (num) => (num * 2) % 10000,
  S: (num) => (num ? num - 1 : 9999),
  L: (num) => (num % 1000) * 10 + ~~(num / 1000),
  R: (num) => (num % 10) * 1000 + ~~(num / 10),
};
const orders = ["D", "S", "L", "R"];
let result = "";
for (let i = 1; i < input.length - 1; i++) {
  const [a, b] = input[i].split(" ").map(Number);
  let q = [];
  let idx = 0;
  q.push([a, ""]);
  visited = new Array(10000);
  const search = (a, b, cur, visited) => {
    for (let i = 0; i < orders.length; i++) {
      let num = cal[orders[i]](a);
      if (num === b) {
        return cur + orders[i];
      }
      if (visited[num]) continue;
      visited[num] = true;
      q.push([num, cur + orders[i]]);
    }
  };
  while (q.length > idx) {
    let [num, cur] = q[idx++];

    let temp = search(num, b, cur, visited);
    if (temp) {
      result += temp + "\n";
      break;
    }
  }
}
console.log(result);

File: DFS&BFS/python/9019.py
from collections import deque
import sys
input = sys.stdin.readline

orders = ["L","R","D","S"]
for _ in range(int(input())):
    visited = [False for _ in range(10000)]
    q = deque()
    n, target = map(int,input().split())
    q.append(['',n])
    visited[n] = True
    check_bool=False
    while q:
        track,num = q.popleft()
        for order in orders:
            if order == "L":
                x = (num%1000)*10 + num//1000
            elif order == "R":
                x = (num%10)*1000 + num//10
            elif order =="D":
                x = (num*2)%10000
            else:
                x = num-1
                if x == -1:
                    x = 9999
            if visited[x]:
                continue
            if x == target:
                print(track+order)
                check_bool=True
                break
            visited[x] = True
            q.append([track+order,x])
        if check_bool:
            break